### DISCUSSION TITLE
request for "Data Graph" with a bigger screen trace

### DISCUSSION THREAD

**Discussion by Breastfed** [collin80/SavvyCAN#279] [Ideas]
Hi All
My first time here and I'm very much of a beginner to RE CAN data, this has become rather interesting as I get further into it .. so be gentle as I'm no programmer LOL.

I have some feature requests based on other ECU recording software i use and find it rather useful.
My laptop has an OLED screen and the high resolution screen is great but hard to see at times ha-ha.

1. Is there any way to make the trace bigger in the Data Graphs ? (or a global trace width setting across all graphs) 

2. Is it possible to invert a data graph some how? I understand this would be done with XOR with FF but I have no idea how to go about this.

3. In the "Frame Data Analysis" some Trace colours are hard to pick out when I try to match them up with which bit number / colour legend. If you could directly "click on" to high lite the trace of interest or even "hover over" it with the mouse and the corresponding "Bit number /colour legend" actively changed  as you move back and forth over the trace.

Regards
Breastfed

**Comment by collin80** [comment]
1. There's a way. I didn't do it because the current graphing system does its best to use OpenGL acceleration. The underlying 3D hardware accelerates single pixel lines but usually not so much thicker lines. So, performance suffers if the line is drawn thicker. This isn't such a big deal if you are graphing 1000 points but if you have a capture with 10 million points then the slowdown becomes much more apparent. I should, perhaps, allow for setting a thicker line width. The performance penalty might not be as large as I had seen in the past either. I haven't tried setting the line width thicker in a few years.
2. I suppose you could set the scaling factor to -1 to invert it. I don't remember if I allowed for negative scaling or not. I don't think it tries to stop you.
3. That's a good idea. Yes, I have trouble differentiating the colors sometimes too. I tried to pick reasonably distinct colors but maybe they aren't distinct enough. The hovering idea might work out pretty well.

**Comment by Breastfed** [comment]
Hi Collin
Appreciate the reply.... looks like its just us at the moment, I've never been the first to join a forum Haha
BTW A scaling factor of -1 worked perfectly.

Another feature request that's very useful
In the Data Graph > right click menu > [Values at the cursor.pdf](https://github.com/collin80/SavvyCAN/files/6002555/Values.at.the.cursor.pdf)
is this possible to add a cursor (1st vertical line) that displays the values that cross this cursor in a table, add a 2nd cursor line so there's a 2nd column of values and the 3rd column in the table displays the difference 


Breastfed

**Comment by collin80** [comment]
I just pushed a commit to github that you might be interested in. In a little bit it'll be automatically compiled to binary form too. Now added are many additional graphing modes. You can set the line width now. You can turn lines off and graph just the points. You can set how points look (whether or not you are also drawing lines in between). You can now set translucent fills on the graphs. I also fixed up loading and saving of graph definitions.

**Comment by Breastfed** [comment]
Hi Collin
I like the line width feature that's perfect.
BTW the Black "fill colour" kills the display performance (HP Envy i7 10th Gen / nvidia GeForce RTX)

is this a Bug?
See the attached regarding the Data graph and the disappearing trace issues
[SavvyCAN V202 GRAPH ISSUES.pdf](https://github.com/collin80/SavvyCAN/files/6010959/SavvyCAN.V202.GRAPH.ISSUES.pdf)
Be nice if there was a no "Fill colour" option or some transparency as the foreground fill colour blocks the background trace ...or I need to adapt my way of thinking to understand this feature.

Breastfed

**Reply by collin80** [reply]
Hmmm, those things just don't happen to me. Try as I might, I can't get the graphs to disappear. I also have an i7 10th gen machine with an nVidia RTX. But, I do most all my testing in Linux. It's possible that there is some sort of display driver issue causing this or that it relates to Windows. I'll test in Windows too. Also, I enabled OpenGL from the main preferences window in SavvyCAN. Have you tried it both with and without OpenGL to see if it makes any difference?

I was actually surprised by how good the performance was for me even with 2-3 graphs all having translucent fill colors. It seems to work OK even with fairly large data sets.

**Comment by Breastfed** [comment]
Hi Collin
I'll tried the OpenGL setting.... still the same.
You mention translucent, but i don't see that as the Grid is not always visible except for when the trace has a dashed line, when you high light a trace I'm guessing you can see the other trace "through" the default black fill ? my fill colours are solid and not translucent.
Could you post a screenshot please.

Is it possible to add a "no fill" option / tick box ?

Breastfed

**Reply by collin80** [reply]
![Screenshot_20210221_210043](https://user-images.githubusercontent.com/2145125/108647982-04cf4980-7488-11eb-879d-31cccda06eee.jpg)

There you can see that I get nice translucent fills. It's plenty speedy too. Yes, when you select a graph it goes to dashed lines and the fill disappears so you can be pretty sure which graph you have selected. There is a no fill option kind of... you see, if you set the translucency of the fill to 0 then it won't try filling. Not exactly a super intuitive approach, I know. That's just how it works with my quick initial attempt. I should figure out something better for that, maybe a button.

**Comment by Breastfed** [comment]
where do i set translucency to 0

**Comment by Breastfed** [comment]
was it the Alpha channel... seems to have done the trick

**Reply by collin80** [reply]
Yes, the alpha channel. Sorry, I forget that not everyone would automatically know that. The alpha channel is transparency and it runs from 0 to 255 in this case. 0 being totally transparent (and thus off) and 255 being totally solid. I have actually never tried the fills totally solid. A value of about 40-120 is plenty.

**Comment by Breastfed** [comment]
Hi Collin
I never knew about the alpha channel setting 
BTW all of my graph definitions are from V199 & V200, it looks like these definitions are forward and backwards compatible with V202. I've created 1 definition from V202 and this also appears to work ok with either V199 & V200 or should I save and use specific definitions for V202?

Breastfed

**Reply by collin80** [reply]
All graph definitions from V199 onward should work with each other. But, the V202 version has more fields. You won't get fill color or line width saved in the older definition format but the older code should still load those definitions fine, it just will ignore the extra info. So, you can load older defs but you'll have to do more set up afterward. You can save in V202 and load in V199 and it'll work but if you save again you'll lose the extra stuff. Still, more or less you can mix and match with whichever version works best. In theory V202 should have everything V199 did so there may be no reason to use the older versions.
